compiler/noirc_frontend/src/monomorphization/tests.rs

```diff
@@ -0,0 +1,37 @@
+#![cfg(test)]
+use crate::tests::get_program;
+
+use super::{ast::Program, errors::MonomorphizationError, monomorphize};
+
+fn get_monomorphized(src: &str) -> Result<Program, MonomorphizationError> {
+    let (_parsed_module, mut context, errors) = get_program(src);
+    assert!(
+        errors.iter().all(|err| !err.is_error()),
+        "Expected monomorphized program to have no errors before monomorphization, but found: {errors:?}"
+    );
+
+    let crate_id = context.root_crate_id();
+    let main = context
+        .get_main_function(crate_id)
+        .unwrap_or_else(|| panic!("get_monomorphized: test program contains no 'main' function"));
+    monomorphize(main, &mut context.def_interner, false)
+}
+
+#[test]
+fn recursive_type_errors() {
+    let src = "
+        fn main() {
+            let _tree: Tree<Tree<Tree<()>>> = Tree::Branch(
```

Comment:
This type may be recursive but it is not infinite. Should we not allow this? This code is legal in Rust.

Reply:
I wanted to allow it but we panic when trying to monomorphize it. I figured we should fix the panic now then could look into relaxing the error and fixing this case later. I'll add a comment.

Reply:
Ok fair enough. Let's capture that we want to relax the error in an issue though.